docs/design/018-new-http-api.md

```diff
@@ -0,0 +1,486 @@
+- `udp`: a message-oriented network socket using the User Datagram Protocol.
+- `ipc`: a mechanism for communicating between processes on the same machine, typically using files.
+
+The Socket state can either be _active_—meaning connected for a TCP socket, bound for a UDP socket, or open for an IPC socket—, or _inactive_—meaning disconnected, unbound, or closed, respectively.
+
+##### Example
+
+- TCP:
+```javascript
+import { TCP } from 'k6/x/net';
+
+export default async function () {
+  const socket = await TCP.open('192.168.1.1:80', {
+                            // default      | possible values
+    ipVersion: 0,           // 0            | 4 (IPv4), 6 (IPv6), 0 (both)
+    keepAlive: true,        // false        |
+    lookup: null,           // dns.lookup() |
+    proxy: 'myproxy:3030',  // ''           |
+  });
+  console.log(socket.active); // true
+
+  // Writing directly to the socket.
+  // Requires TextEncoder implementation, otherwise typed arrays can be used as well.
+  await socket.write(new TextEncoder().encode('GET / HTTP/1.1\r\n\r\n'));
+
+  // And reading...
+  socket.on('data', (data) => {
+    console.log(`received ${data}`);
+    socket.close();
+  });
+
+  await socket.done();
+}
+```
+
+- UDP:
+```javascript
+import { UDP } from 'k6/x/net';
+
+export default async function () {
+  const socket = new UDP.open('192.168.1.1:9090');
+
+  await socket.write(new TextEncoder().encode('GET / HTTP/1.1\r\n\r\n'));
+}
+```
+
+- IPC:
+```javascript
+import { IPC } from 'k6/x/net';
+import { Client } from 'k6/x/net/http';
+
+export default async function () {
+  const socket = await IPC.open('/tmp/unix.sock');
+
+  console.log(socket.file.path); // /tmp/unix.sock
+
+  // The HTTP client supports communicating over a Unix socket.
+  const client = new Client({
+    socket: socket,
+  });
+  await client.get('http://unix/get');
+}
+```
+
+#### Client
+
+An HTTP Client is used to communicate with an HTTP server.
+
+##### Examples
+
+- Using a client with default transport settings, and making a GET request:
+```javascript
+import { Client } from 'k6/x/net/http';
+
+export default async function () {
+  const client = new Client();
+  const response = await client.get('https://httpbin.test.k6.io/get');
+  const jsonData = await response.json();
+  console.log(jsonData);
+}
+```
+
+- Passing a socket with custom transport settings, some HTTP options, and making a POST request:
+```javascript
+import { TCP } from 'k6/x/net';
+import { Client } from 'k6/x/net/http';
+
+export default async function () {
+  const socket = await TCP.open('10.0.0.10:80', { keepAlive: true });
+  const client = new Client({
+    socket: socket,
+    proxy: 'https://myproxy',
+    version: 1.1,                     // force a specific HTTP version
+    headers: { 'User-Agent': 'k6' },  // set some global headers
+  });
+  await client.post('http://10.0.0.10/post', {
+    json: { name: 'k6' }, // automatically adds 'Content-Type: application/json' header
+  });
```

Comment:
Same logic as https://github.com/grafana/k6/pull/2971/files#r1161730930 - we shouldn't add any special handling of `json` options in the default Client. This is not composable, it's the opposite of that - trying to make a `Client` that satisfies all use cases. And it adds "automagic behavior", as your `// automatically adds 'Content-Type: application/json' header` comment directly shows.

The request bodies that the normal HTTP Client accepts should probably be limited to `string`, `ArrayBuffer` or some sort of a Stream. Anything else needs to be an error. We can have a separate `RESTClient` or `JSONClient` that has additional automated handling of JSON requests and responses.

Reply:
The problem is that it makes the API have weird edge cases. For example, what happens if I provide both `body` and `json` and `formData`? :sweat_smile: Now the library needs to have some extra internal logic for the order of precedence, and we have do document and test that, etc. Having more than one way to do the same thing in the same object is generally not great API design.

Reply:
Actually, from looking at the request library API you linked to, its `json` property is a boolean one, you still have to supply to `body`, right? Which may be even worse than this suggestion, since you now have extra parameters that just control the behavior of how the `body` is automagically processed.